libcaf_net/tests/drivers/autobahn.cpp

```diff
@@ -0,0 +1,130 @@
+public:
+  static auto make() {
+    return std::make_unique<web_socket_app>();
+  }
+
+  caf::error start(caf::net::web_socket::lower_layer* ll) override {
+    down = ll;
+    down->request_messages();
+    return caf::none;
+  }
+
+  caf::error accept(const caf::net::http::request_header&) override {
+    return caf::none;
+  }
+
+  void prepare_send() override {
+    // nop
+  }
+
+  bool done_sending() override {
+    return true;
+  }
+
+  void abort(const caf::error& reason) override {
+    CAF_LOG_ERROR(reason);
+  }
+
+  ptrdiff_t consume_text(std::string_view text) override {
+    down->begin_text_message();
+    auto& buffer = down->text_message_buffer();
+    buffer.insert(buffer.end(), text.begin(), text.end());
+    down->end_text_message();
+    return static_cast<ptrdiff_t>(text.size());
+  }
+
+  ptrdiff_t consume_binary(caf::byte_span bytes) override {
+    down->begin_binary_message();
+    auto& buffer = down->binary_message_buffer();
+    buffer.insert(buffer.end(), bytes.begin(), bytes.end());
+    down->end_binary_message();
+    return static_cast<ptrdiff_t>(bytes.size());
+  }
+
+  caf::net::web_socket::lower_layer* down = nullptr;
+};
+
+int caf_main(caf::actor_system& sys, const config& cfg) {
+  // Read the configuration.
+  auto port = caf::get_or(cfg, "port", default_port);
+  auto pem = ssl::format::pem;
+  auto key_file = caf::get_as<std::string>(cfg, "tls.key-file");
+  auto cert_file = caf::get_as<std::string>(cfg, "tls.cert-file");
+  auto max_connections = caf::get_or(cfg, "max-connections",
+                                     default_max_connections);
+  if (!key_file != !cert_file) {
+    std::cerr << "*** inconsistent TLS config: declare neither file or both\n";
+    return EXIT_FAILURE;
+  }
+  // Open up a TCP port for incoming connections and start the server.
+  auto server
+    = ws::with(sys)
+        // Optionally enable TLS.
+        .context(ssl::context::enable(key_file && cert_file)
+                   .and_then(ssl::emplace_server(ssl::tls::v1_2))
+                   .and_then(ssl::use_private_key_file(key_file, pem))
+                   .and_then(ssl::use_certificate_file(cert_file, pem)))
+        // Bind to the user-defined port.
+        .accept(port)
+        // Limit how many clients may be connected at any given time.
+        .max_connections(max_connections)
+        // Add handler for incoming connections.
+        .on_request([](ws::acceptor<>& acc) {
+          // Ignore all header fields and accept the connection.
+          acc.accept();
+        })
+        // When started, run our worker actor to handle incoming connections.
+        .start([] { return web_socket_app::make(); });
+  // Report any error to the user.
+  if (!server) {
+    std::cerr << "*** unable to run at port " << port << ": "
+              << to_string(server.error()) << '\n';
+    return EXIT_FAILURE;
+  }
+  while (server->valid()) {
+    std::this_thread::sleep_for(1s);
+  }
```

Comment:
Not particularly happy with this solution. We can leave it like this for now, but we should provide an API to keep servers alive. Even if it's just something like `sys.wait_for(server)`.